synapse/handlers/message.py

```diff
@@ -225,6 +243,97 @@ def get_joined_members(self, requester, room_id):
             for user_id, profile in iteritems(users_with_profile)
         }
 
+    def maybe_schedule_expiry(self, event):
+        """Schedule the expiry of an event if there's not already one scheduled,
+        or if the one running is for an event that will expire after the provided
+        timestamp.
+
+        Args:
+            event (EventBase): The event to schedule the expiry of.
+        """
+        expiry_ts = event.content.get(EventContentFields.SELF_DESTRUCT_AFTER)
+        if not isinstance(expiry_ts, int) or event.is_state():
+            return
+
+        if self._scheduled_expiry:
```

Comment:
we don't know at this point if the task has run or not, so there's a race:

* expiry starts running, and gets as far as starting to run get_next_event_to_expire, which decides the next expiry is in one week
* before the expiry task is rescheduled, `maybe_schedule_expiry` is called for an event which expires in 5m.
* but `maybe_schedule_expiry` sees that _scheduled_expiry is truthy, and the scheduled time is in the past (we know this is the case, since it has been run), so does nothing
* so the 5m expiry is lost.

You can solve this problem by checking `_scheduled_expiry.active` or (as I am suggesting) setting _scheduled_expiry to None as soon as the expiry task starts.